text/0000-inline-asm.md

```diff
@@ -0,0 +1,1055 @@
+
+These allow the compiler to better optimize code using `asm!`, for example by eliminating pure `asm!` blocks whose outputs are not needed.
+
+See the reference for the full list of available options and their effects.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Inline assembler is implemented as an unsafe macro `asm!()`.
+The first argument to this macro is a template string literal used to build the final assembly.
+The following arguments specify input and output operands.
+When required, options are specified as the final argument.
+
+The following ABNF specifies the general syntax:
+
+```
+dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
+reg_spec := <register class> / "<explicit register>"
+operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
+reg_operand := dir_spec "(" reg_spec ")" operand_expr
+operand := reg_operand / "const" const_expr / "sym" path
+option := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
+options := "options(" option *["," option] [","] ")"
+asm := "asm!(" format_string *("," [ident "="] operand) ["," options] [","] ")"
+```
+
+The macro will initially be supported only on ARM, AArch64, x86, x86-64 and RISC-V targets. Support for more targets may be added in the future. The compiler will emit an error if `asm!` is used on an unsupported target.
+
+[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax
+
+## Template string
+
+The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.
+
+As with format strings, named arguments must appear after positional arguments. Explicit register operands must appear at the end of the operand list, after any named arguments if any. Explicit register operands cannot be used by placeholders in the template string. All other operands must appear at least once in the template string, otherwise a compiler error is generated.
+
+The exact assembly code syntax is target-specific and opaque to the compiler except for the way operands are substituted into the template string to form the code passed to the assembler.
+
+The 4 targets specified in this RFC (x86, ARM, AArch64, RISC-V) all use the assembly code syntax of the GNU assembler (GAS). On x86, the `.intel_syntax noprefix` mode of GAS is used. On ARM, the `.syntax unified` mode is used. These targets impose an additional restriction on the assembly code: any assembler state (e.g. the current section which can be changed with `.section`) must be restored to its original value at the end of the asm string. Assembly code that does not conform to the GAS syntax will result in assembler-specific behavior.
+
+[rfc-2795]: https://github.com/rust-lang/rfcs/pull/2795
+
+## Operand type
```

Comment:
Only `inout` requires a mutable place. I will adjust the wording.

Reply:
Wouldn't `out` require it too if the place is already (potentially) initialized?

Reply:
No, this code is perfectly valid:
```rust
let foo: u32;
asm!("...", out(reg) foo);
```

Reply:
Overwriting an already initialized immutable place is of course not allowed.

Reply:
So, from move/borrow checker point of view `in*` operands consume the expression, and `*out` operands assign to the expression.

Then all the initialization/mutability properties follow from that.

Reply:
Hmm. But there is also `clobber` to consider, which assigns to the expression *without* giving it a well-defined value.

That is, if I understand correctly:

```rust
let foo: u32;
asm!("...", out(reg) foo);
a_use_of(foo); // legal
```

```rust
let foo: u32;
asm!("...", clobber(reg) foo);
a_use_of(foo); // illegal
```

(also, does the current system have a notion of `inclobber`, which is like `inout` in that it only clobbers the variable after it has read the relevant data from it? Does it need such a notion?)


I don't remember if the current move/borrow-checker, liveness, and variable-initialization analyses currently have a way to express the `clobber` effect, but even if they don't, I cannot imagine it would be terribly hard to add.

----

In any case, it would be good to spell this out explicitly in the RFC, since I was looking for it as I was reading and could not find it.

Reply:
There is no `clobber` operand type. Instead you can specify that a register is clobbered by declaring it as an output which is discarded:
```rust
let foo: u32 = 3;
asm!("...", out(reg) _);
asm!("...", inout(reg) foo => _);
```

Reply:
Oh I'm sorry; I somehow managed to read the section about clobbering without noting the general syntax you have chosen for it.